### PR TITLE
Add mdarray

### PR DESCRIPTION
**Pull request by crtrott**

### CODE REVIEW

File: include/experimental/__p1684_bits/mdarray.hpp [hunk 1]
```diff
@@ -0,0 +1,484 @@
...
+  MDSPAN_INLINE_FUNCTION
+  constexpr mdarray(const mdarray<OtherElementType, OtherExtents, OtherLayoutPolicy, OtherContainer>& other)
+    : map_(other.mapping()), ctr_(other.container())
+  {
+    static_assert(is_constructible<extents_type, OtherExtents>::value, "");
+  }
+
+  // Constructors for container types constructible from a size and allocator
+  MDSPAN_TEMPLATE_REQUIRES(
+    class Alloc,
```
Comment (AnabelSMRuggiero) [inline include/experimental/__p1684_bits/mdarray.hpp:252]: As is, I suspect these constructors are under constrained. Constructing an `mdarray<int, ..., std::vector<int>>` with an `extents_type` and a `float` (as an example), would satisfy the requirements of this constructor. That causes the resulting vector to have the appropriate number of elements, each initialized the the value of the (truncated) `float`. Not the worst thing to possibly happen here, but I've not exhaustively checked all constructors.

**Review by AnabelSMRuggiero — APPROVED** [review]
Other than my comment about under-constrained, but not invariant violating allocator constructors, nothing else seems out of line with the current proposal to me.

**Comment by rscohn2** [comment]
Where can I see a spec for the proposal?

**Comment by rscohn2** [comment]
> Where can I see a spec for the proposal?

Found it: http://www7.open-std.org/jtc1/sc22/wg21/docs/papers/2022/p1684r1.html

File: include/experimental/__p1684_bits/mdarray.hpp (hunk 1 above)
Comment (AnabelSMRuggiero) [reply include/experimental/__p1684_bits/mdarray.hpp:252]: It occurred to me today `requires uses_allocator_v<container_type, Alloc>` or equivalent for pre-C++20 should be sufficient for constraining the allocator constructors.

**Comment by mhoemmen** [comment]
@rscohn2 One can use the redirect link https://wg21.link/p1684 to find the latest published version of a proposal.

**Review by mhoemmen — APPROVED** [review]
Approving because this is already an improvement; we can come back to other issues later : - )

Excellent test using PMR btw!

**Pull request merged**